python/ray/tempfile_services.py

```diff
@@ -0,0 +1,242 @@
+    # Change the default root by providing the directory named by the
+    # TMPDIR environment variable.
+    temp_dir = tempfile.TemporaryDirectory(
+        prefix='ray_{pid}_'.format(pid=os.getpid()))
+    temp_root = temp_dir.name
+
+
+def get_logs_dir_path():
+    """Get a temp dir for logging."""
+    logs_dir = os.path.join(temp_root, 'logs')
+    try_to_create_directory(logs_dir)
+    return logs_dir
+
+
+def get_rl_dir_path():
+    """Get a temp dir that will be used by some of the RL algorithms."""
+    rl_dir = os.path.join(temp_root, 'ray')
+    try_to_create_directory(rl_dir)
+    return rl_dir
+
+
+def get_sockets_dir_path():
+    """Get a temp dir for sockets"""
+    sockets_dir = os.path.join(temp_root, 'sockets')
+    try_to_create_directory(sockets_dir)
+    return sockets_dir
+
+
+def get_raylet_socket_name(suffix=None):
+    """Get a socket name for raylet.
+
+    This function could be unsafe. The socket name may
+    refer to a file that did not exist at some point, but by the time
+    you get around to creating it, someone else may have beaten you to
+    the punch.
+    """
+    sockets_dir = get_sockets_dir_path()
+
+    if suffix is None:
+        raylet_socket_name = tempfile.mktemp(prefix='raylet_', dir=sockets_dir)
+    else:
+        raylet_socket_name = os.path.join(sockets_dir,
+                                          'raylet_{}'.format(suffix))
+    return raylet_socket_name
+
+
+def get_object_store_socket_name():
+    sockets_dir = get_sockets_dir_path()
+    if RAY_OBJECT_STORE_SOCKET_NAME in os.environ:
+        return os.path.join(sockets_dir,
+                            os.environ[RAY_OBJECT_STORE_SOCKET_NAME])
+    else:
+        return tempfile.mktemp(prefix='plasma_store_', dir=sockets_dir)
+
+
+def get_plasma_manager_socket_name():
+    sockets_dir = get_sockets_dir_path()
+    return tempfile.mktemp(prefix='plasma_manager_', dir=sockets_dir)
+
+
+def get_local_scheduler_socket_name(suffix=None):
+    """Get a socket name for local scheduler.
+
+    This function could be unsafe. The socket name may
+    refer to a file that did not exist at some point, but by the time
+    you get around to creating it, someone else may have beaten you to
+    the punch.
+    """
+    sockets_dir = get_sockets_dir_path()
+
+    if suffix is None:
+        raylet_socket_name = tempfile.mktemp(prefix='scheduler_',
+                                             dir=sockets_dir)
+    else:
+        raylet_socket_name = os.path.join(sockets_dir,
+                                          'scheduler_{}'.format(suffix))
+    return raylet_socket_name
+
+
+def get_random_ipython_notebook_path(port):
+    """Get a new random ipython notebook path"""
+
+    notebook_filepath = os.path.join(
+        os.path.dirname(os.path.abspath(__file__)), "WebUI.ipynb")
+    # We copy the notebook file so that the original doesn't get modified by
+    # the user.
+    notebook_name = tempfile.mktemp(suffix='.ipynb', prefix='ray_ui_',
```

Comment:
`mktemp` is deprecated.

Reply:
I think the question is the approach of generating a filename is a good one. I think it is not because it cannot assure unique names. So I would argue that Ray should be changed so that it never uses temporary filename, but uses file objects directly. Could fix #2594.